core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleMapOutputWriter.java

```diff
@@ -154,42 +154,42 @@ private DefaultShufflePartitionWriter(int partitionId) {
     }
 
     @Override
-    public OutputStream openStream() throws IOException {
+    public OutputStream toStream() throws IOException {
       initStream();
       stream = new PartitionWriterStream();
       return stream;
     }
 
     @Override
-    public long closeAndGetLength() {
+    public FileChannel toChannel() throws IOException {
+      initChannel();
+      currChannelPosition = outputFileChannel.position();
+      return outputFileChannel;
+    }
+
+    @Override
+    public long getNumBytesWritten() {
       if (outputFileChannel != null && stream == null) {
         try {
           long newPosition = outputFileChannel.position();
-          long length = newPosition - currChannelPosition;
-          partitionLengths[partitionId] = length;
-          currChannelPosition = newPosition;
-          return length;
+          return newPosition - currChannelPosition;
         } catch (Exception e) {
-          log.error("The currPartition is: " + partitionId, e);
-          throw new IllegalStateException("Attempting to calculate position of file channel", e);
+          log.error("The currPartition is: {}", partitionId, e);
+          throw new IllegalStateException("Failed to calculate position of file channel", e);
         }
+      } else if (stream != null) {
+        return stream.getCount();
       } else {
-        try {
-          stream.close();
-        } catch (Exception e) {
-          throw new IllegalStateException("Attempting to close output stream", e);
-        }
-        int length = stream.getCount();
-        partitionLengths[partitionId] = length;
-        return length;
+        return 0;
       }
     }
 
     @Override
-    public FileChannel openChannel() throws IOException {
-      initChannel();
-      currChannelPosition = outputFileChannel.position();
-      return outputFileChannel;
+    public void close() throws IOException {
```

Comment:
I think it might be good to keep track of the `outputFileChannel` close as well. Basically, we should ensure that this method is called before `getNumBytesWrittern()` is called. Otherwise, I might actually be in favor of having `close()` return the long to ensure that `getNumBytesWritten()` can never be called before closing.